src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+                }
+                catch (ArrayTypeMismatchException)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidArrayType, nameof(array));
+                }
+            }
+
+            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>, IEnumerator
+            {
+                private readonly PriorityQueue<TElement, TPriority> _queue;
+                private readonly int _version;
+
+                private int _index;
+                private (TElement element, TPriority priority)? _currentElement;
+
+                private const int FirstCallToEnumerator = -2;
+                private const int EndOfEnumeration = -1;
+
+                internal Enumerator(PriorityQueue<TElement, TPriority> queue)
+                {
+                    _queue = queue;
+                    _version = queue._version;
+                    _index = FirstCallToEnumerator;
+                    _currentElement = default;
+                }
+
+                public void Dispose()
+                {
+                    _index = EndOfEnumeration;
+                }
+
+                public bool MoveNext()
+                {
+                    bool advancedEnumerator;
+
+                    if (_version != _queue._version)
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_EnumFailedVersion);
+                    }
+
+                    if (_index == FirstCallToEnumerator)
+                    {
+                        _index = 0;
+                        advancedEnumerator = (_queue._size > 0);
+
+                        if (advancedEnumerator)
+                        {
+                            _currentElement = _queue._nodes[_index];
+                        }
+                        else
+                        {
+                            _index = EndOfEnumeration;
+                        }
+
+                        return advancedEnumerator;
+                    }
+
+                    if (_index == EndOfEnumeration)
+                    {
+                        return false;
+                    }
+
+                    _index++;
+                    advancedEnumerator = (_index < _queue._size);
+
+                    if (advancedEnumerator)
+                    {
+                        _currentElement = _queue._nodes[_index];
+                    }
+                    else
+                    {
+                        _currentElement = default;
+                    }
+
+                    return advancedEnumerator;
+                }
+
+                public (TElement element, TPriority priority) Current
+                {
+                    get
+                    {
+                        if (_index < 0)
+                        {
+                            ThrowEnumerationNotStartedOrEnded();
+                        }
+                        return _currentElement!.Value;
```

Comment:
Shouldn't this return just _currentElement?

Reply:
Oh, I see it's a Nullable<T>. In that case, I don't think you need the `!`.
@stephentoub should the compiler complain or am I reading wrong?

Reply:
> should the compiler complain

Yes, it factors in `Nullable<T>`.  I expect the warning is something along the lines of "Nullable value type may be null".

Reply:
ah yes

Reply:
I think we could avoid making `_currentElement` nullable since we use `_index` to determine whether it is populated or not.